client/cache.go

```diff
@@ -0,0 +1,292 @@
+func (o *objectCache) CacheObject(
+	gvk schema.GroupVersionKind, namespace string, name string, object *unstructured.Unstructured,
+) {
+	objID := ObjectIdentifier{
+		Group:     gvk.Group,
+		Version:   gvk.Version,
+		Kind:      gvk.Kind,
+		Namespace: namespace,
+		Name:      name,
+	}
+
+	o.CacheFromObjectIdentifier(objID, []unstructured.Unstructured{*object})
+}
+
+// CacheFromObjectIdentifier will cache a list of objects for the input object identifier.
+func (o *objectCache) CacheFromObjectIdentifier(objID ObjectIdentifier, objects []unstructured.Unstructured) {
+	o.cache.Store(objID, objects)
+}
+
+// UncacheObject will entirely remove the cache entry of the object.
+func (o *objectCache) UncacheObject(gvk schema.GroupVersionKind, namespace string, name string) {
+	objID := ObjectIdentifier{
+		Group:     gvk.Group,
+		Version:   gvk.Version,
+		Kind:      gvk.Kind,
+		Namespace: namespace,
+		Name:      name,
+	}
+
+	o.UncacheFromObjectIdentifier(objID)
+}
+
+// UncacheObject will entirely remove the cache entries for the list query.
+func (o *objectCache) UncacheList(gvk schema.GroupVersionKind, namespace string, selector labels.Selector) {
+	if selector == nil {
+		selector = labels.NewSelector()
+	}
+
+	objID := ObjectIdentifier{
+		Group:     gvk.Group,
+		Version:   gvk.Version,
+		Kind:      gvk.Kind,
+		Namespace: namespace,
+		Selector:  selector.String(),
+	}
+
+	o.UncacheFromObjectIdentifier(objID)
+}
+
+// UncacheObject will entirely remove the cache entries for the object identifier.
+func (o *objectCache) UncacheFromObjectIdentifier(objID ObjectIdentifier) {
+	o.cache.Delete(objID)
+}
+
+// GVKToGVR will convert a GVK to a GVR and cache the result for 10 minutes if found, and 1 minte if not found.
+func (o *objectCache) GVKToGVR(gvk schema.GroupVersionKind) (ScopedGVR, error) {
+	lockedGVRObj := &lockedGVR{lock: &sync.RWMutex{}, expires: time.Now().Add(10 * time.Minute)}
+	lockedGVRObj.lock.Lock()
+
+	loadedLockedGVRObj, loaded := o.gvkToGVRCache.LoadOrStore(gvk, lockedGVRObj)
+	if loaded {
+		// If the value was loaded (not stored), there means there is a cached value or it is being populated.
+		lockedGVRObj = loadedLockedGVRObj.(*lockedGVR)
+		lockedGVRObj.lock.RLock()
+		defer lockedGVRObj.lock.RUnlock()
+
+		if !lockedGVRObj.isExpired() {
+			// If stored but nil, this means the previous call failed.
+			if lockedGVRObj.gvr == nil {
+				return ScopedGVR{}, ErrNoVersionedResource
+			}
+
+			return *lockedGVRObj.gvr, nil
+		}
+
+		// The cache is expired, get a write lock to update the entry.
+		lockedGVRObj.lock.Lock()
+	}
+
+	defer lockedGVRObj.lock.Unlock()
+
+	groupVersion := gvk.GroupVersion().String()
+
+	resources, err := o.discoveryClient.ServerResourcesForGroupVersion(groupVersion)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			klog.Infof("The group version was not found: %s", groupVersion)
+
+			// Reduce the cache time to 1 minute from the request if it's from the resource not being present on the
```

Comment:
Why it is one minute? Is there any reason? If a resource is not found, that means.. it is removed from kube.

Reply:
As discussed in our meeting, the idea is so that if a CRD is missing, it doesn't cause multiple Kubernetes API queries every time there is a template referencing this resource. We wait a minute to try again to see if the CRD is installed.